modules/play-json/src/main/scala/PlayJsonModule.scala

```diff
@@ -0,0 +1,170 @@
+
+      def apply[A](fschema: FSchema[R.Prim, R.SumTermId, R.ProductTermId, A]): LabelledSchema[A] =
+        (label, fschema)
+    }
+
+  final val labelRecordFields: HCoalgebra[HEnvT[Boolean, RSchema, ?[_], ?], LabelledSchema] =
+    new (LabelledSchema ~> HEnvT[Boolean, RSchema, LabelledSchema, ?]) {
+
+      def apply[A](seed: LabelledSchema[A]): HEnvT[Boolean, RSchema, LabelledSchema, A] =
+        seed match {
+          case (x, Fix(r @ Record(_, _))) => HEnvT(x, r.hmap(ascribeWith(true)))
+          case (x, Fix(:*:(left, right))) =>
+            HEnvT(
+              x,
+              :*:(
+                ascribeWith(false)(left),
+                ascribeWith(x)(right)
+              )
+            )
+          case (x, s) => HEnvT(x, s.unFix.hmap(ascribeWith(x)))
+        }
+    }
+
+  // This is needed in order to use `traverse` in `reads`.
+  implicit private val readsApplicative = new Applicative[JsResult] {
+
+    def point[A](a: => A): JsResult[A] = JsSuccess(a)
+
+    def ap[A, B](fa: => JsResult[A])(f: => JsResult[A => B]): JsResult[B] = (f, fa) match {
+      // Shamelessly copied from the play-json library
+      case (JsSuccess(f, _), JsSuccess(a, _)) => JsSuccess(f(a))
+      case (JsError(e1), JsError(e2))         => JsError(JsError.merge(e1, e2))
+      case (JsError(e), _)                    => JsError(e)
+      case (_, JsError(e))                    => JsError(e)
+
+    }
+  }
+
+  // This is needed to allow undefined optional fields to be treated as `None`.
+  final private def undefinedAsNull[A](field: String, r: Reads[A]): Reads[A] = Reads { json =>
+    ((json \ field) match {
+      case JsDefined(v) => r.reads(v)
+      case _            => r.reads(JsNull)
+    }).repath(JsPath \ field)
+  }
+
+  final private val labellingSeed =
+    new (FSchema[R.Prim, R.SumTermId, R.ProductTermId, ?] ~> LabelledSchema) {
+
+      def apply[A](fSchema: FSchema[R.Prim, R.SumTermId, R.ProductTermId, A]): LabelledSchema[A] =
+        (false, fSchema)
+    }
+
+  implicit final def reads(
+    implicit primNT: R.Prim ~> Reads,
+    branchLabel: R.SumTermId <~< String,
+    fieldLabel: R.ProductTermId <~< String
+  ): RInterpreter[Reads] =
+    new Interpreter[R.Prim, R.SumTermId, R.ProductTermId, Reads] {
+
+      val alg = new (HEnvT[Boolean, RSchema, Reads, ?] ~> Reads) {
+
+        def apply[A](schema: HEnvT[Boolean, RSchema, Reads, A]): Reads[A] = schema.fa match {
+          case One() =>
+            Reads {
+              case JsNull => JsSuccess(())
+              case _      => JsError(Seq(JsPath -> Seq(JsonValidationError("error.expected.null"))))
+            }
+          case :+:(left, right) =>
+            Reads(
+              json =>
+                left
+                  .reads(json)
+                  .fold(
+                    el =>
+                      right.reads(json).map(\/-.apply) match {
+                        case JsError(er) => JsError(JsError.merge(el, er))
```

Comment:
Why merge the errors of both alternatives?

Given the `Schema[Role]` defined in the tests, I would expect the following behavior from decoders:

- `{ "user": {} }` would be decoded as a failure with the following (single) error: “Missing field ’active’” ;
- `{ "moderator": {} }` would be decoded as a failure with the following error: “Unknown alternative label: 'moderator'. Possible values are: 'user', 'admin'.” ;
- `{}` would be decoded as a failure with the following error: “Missing alternative label”.

Reply:
I agree that error management isn't completely satisfactory.

This is due to:
1. the fact that (as you pointed out in another comment) we don't leverage our encoding for unions, which wraps branches in an object with a single field whose name is the branch's label (mostly because we're unable to do so ATM)
2. the semantics of `JsError`. Getting a `JsError(Seq(x, y, z))` means "My unfulfilled expectations are `x` **and** `y` **and** `z`". But in the case of a `union`, we would need a "or" semantic.

But know that I've written it down, I realise that 2. is a consequence of 1. If we find a way to properly leverage our encoding for branches, we should be able to get better error messages.

I'm not sure we can/want to satisfy your second point though (detecting "unknown branches")

Reply:
> I'm not sure we can/want to satisfy your second point though (detecting "unknown branches")

At least, just reporting “unknown alternative: 'moderator'”, would be fine.

Reply:
That might be doable.